### ISSUE TITLE
Gatsby 3 syntax support

### ISSUE THREAD

**Issue by mosesoak** [mrmckeb/typescript-plugin-css-modules#129]
**Is your feature request related to a problem? Please describe.**

Yes. I'm unable to use your solution with Gatsby 3, which supports:
`import { classOne, classTwo } from './styles.module.scss';`

and supports but doesn't recommend,

`import * as styles from './styles.module.scss';`

but no longer supports:

`import styles from './styles.module.scss';`

At best, after installing the plugin, setting up VS Code as recommended and reloading the project, I've simply gotten `string` types to show up by doing `export = classes;` in the d.ts file, `export default classes;` fails entirely.

**Describe the solution you'd like**

Ideally I'd like to use Gatsby 3 with its standard `gatsby-plugin-sass` and scss modules, and be able to both get type hints in VS Code and ideally be able to jump to the definition of the style in the scss file from the component file.

Even more ideal than even this language plugin would be a way to just make this easier for anyone who wants to use Gatsby and scss modules work with Typescript out of the box in VS Code. For example could someone author a VS Code plugin that would solve this? Should we be petitioning Microsoft to somehow make this easier? It's frustrating that such a fundamental technology (css/scss modules + import) is so lacking in official support.

**Describe alternatives you've considered**

Currently the only solution that partially works is another 3rd-party lib, `gatsby-plugin-scss-typescript`, which generates individual d.ts files for each scss file. However, that adds quite a bit of clutter to the project, and clicking through does not jump into the class definition, just into the generated type. That said, it successfully adds code hinting.

**Additional context**

I'm on a PC running VS Code with remote-WSL. While testing out this solution I upgraded to TS 4.3.2.

**Comment by mosesoak** [comment]
I found a VS Code plugin called [CSS Modules](https://github.com/clinyong/vscode-css-modules) that currently does work to provide code hinting and jump-to-definition. It doesn't support absolute paths, curly-brace module imports, and doesn't show any errors on incorrect class names, but so far it's the closest thing I've found to something that works with `import * as styles` syntax.

**Comment by mrmckeb** [comment]
Hi @mosesoak, thanks for all of the detail above.

I've just tested this locally and can confirm that it does work with the first example, and not the second.

You can see this working in VSCode here:
![image](https://user-images.githubusercontent.com/5043083/120072003-49c3d700-c09a-11eb-8e85-a76a380107ac.png)

Just to clarify, are you saying that you can't use this approach and need to use `* as styles`? Or that you can't get the first example to work either? If it's the latter, could you provide a repro?

---

On your other comment, I agree and also wish this was easier. The TypeScript team haven't spent a lot of time on the plugin ecosystem and there are still limitations, there are a few issues on the topic and I know that they're listening - I think it's just prioritisation.

The VSCode plugin is a good alternative to this plugin. The original idea was that this plugin would give build-time errors too, but that's still not possible with plain TypeScript.

**Comment by mosesoak** [comment]
Thanks for your reply. 

Since Gatsby 3 no longer supports `import styles` I need support for ` import * as styles`. The destructured import syntax didn't work for me.

￼I guess I'll be using the VS code extension for now, but was hoping I could use your plugin in tandem to provide type-checking.￼ 

(So far your plug-in doesn't work at all for me with Gatsby 3 -- if you have time, you could spin up a Gatsby project with Typescript and try adding your plugin to cross-test)

**Comment by cedeber** [comment]
Hello, the CSS tree shaking of Parcel v2 is also only available via `import * as styles` or named import. (although it still support the default import)

So I agree that supporting the destructuring import would be nice :-)
Thanks.

**Comment by mrmckeb** [comment]
Hi @mosesoak and @cedeber, I made some changes in the last update (released last weekend and that seems to have fixed `import * as styles` in my testing.

If this syntax, or destructured imports still aren't working for you, it's very likely an issue with the CSS files themselves and how this plugin parses them (maybe a missing setting, or something we can add to help). Providing a reproduction would go a long way in helping me to resolve that type of issue - I'd love to help get this sorted for you!

As an example, you can see here that I have autocompletion/types.
![image](https://user-images.githubusercontent.com/5043083/120886896-e3d6d280-c5f8-11eb-98ad-d31728e36c11.png)

**Comment by mrmckeb** [comment]
Another thought, if your classnames have hyphens, they obviously won't be able to be destructured (just in case).

**Comment by cedeber** [comment]
I upgrade to 3.4.0 and, indeed, it works better now.
But a class named `my_css_class_name` with the setup `"classnameTransform": "asIs"` doesn't work. I believe underscores are valid, aren't they?

**Comment by mrmckeb** [comment]
Hello, they definitely are supported! 😊

I've just double-checked with my test project, and this is what I see with that setting (note the last classname).
![image](https://user-images.githubusercontent.com/5043083/121043209-17625a00-c7bd-11eb-9644-741a3fa84108.png)

There might be something else happening, I'd be happy to take a look if you can provide a minimal reproduction?

**Comment by cedeber** [comment]
Yeah, but does `styles.App_logo` work too? I mean underscores are valid in objects prop names.

**Comment by mrmckeb** [comment]
Yes, it does :)
![image](https://user-images.githubusercontent.com/5043083/121208070-763dd700-c882-11eb-9e30-d59e6e8339e8.png)

**Comment by mosesoak** [comment]
@mrmckeb Thanks for the updates! I have some pretty good results:

- Both * as styles and destructured imports and code-hinting all work now!
- You were right that a line in my scss file was choking the plugin: @import '~/styles/variables'; -- it's the absolute path configuration we use. Changing to a relative import path fixes the error.
- Jump-to-definition does jump to the correct file but does not jump the cursor to the correct class. That's something that does work perfectly with the vs code css modules plugin, which I had deactivated for this test. If I reactivate it, two definitions are shown, theirs correctly shows the class, yours shows filename.scss:238:13 (the file only has 121 lines).

Anyway good progress, thank you for your quick response.

Sidebar: I've been surprised that CSS modules are not really in demand, I sort of thought or hoped that they'd be more embraced as a new standard since I kind of prefer SCSS to css-in-js solutions, which tend to add a new paradigm of their own that just doesn't feel like it will ever become a real standard way of doing things. The Gatsby team wrote back on my question about this saying there's just not much demand for CSS modules & typescript out there. Kind of a bummer. We're continuing to try out different things in our projects, styled-jsx, styled-components, emotion, and frameworks like tailwind. Cool that folks are working on this but I guess styling is just a neverending eddy in the fast-moving world of web dev.

**Comment by cedeber** [comment]
Wow, that's weird on my side. This is how it looks like on intelliJ:

With array notation:
![css_module_1](https://user-images.githubusercontent.com/3330814/121328689-112ac580-c904-11eb-8f9e-ecfc16c75aed.png)

With dot notation:
![css_module_2](https://user-images.githubusercontent.com/3330814/121328742-1f78e180-c904-11eb-8efa-12cd3d79d255.png)

I only have access to `style.default.axis_container`?

The SCSS file only contains this class and I import it via `import * as style from "./styles/Whatever.module.scss"`;

**Comment by mrmckeb** [comment]
Hi @mosesoak, thanks for the kind words!

I think the problem in the ecosystem that everyone has a different config - Less/Sass/SCSS, different ways of aliasing/importing, etc. Because there isn't a standard, no tool will be perfect.

I'm on the team for Create React App and that's one place where I can guarantee certainty ;) I also need to do some spot checking against the default support in Next.js too, but I don't expect issues.


On definitions... I actually had an idea of how to make jump-to-definition work, and it looks like the VSCode extension you mentioned uses that exact approach:
https://github.com/clinyong/vscode-css-modules/blob/master/src/DefinitionProvider.ts

Note that this probably fails in some cases, like when classes are created with concatenation, etc.

TLDR; The final definition file will look very different to the source, and that's why the lines are messed up. We'd need to "reconstruct" the `.d.ts` to have the same line positions for this to work.
```
// Source (your *.css)

.class-b {
   color: red;
   background: blue
}

.class-b {
   color: green;
}

// Definition (virtual *.d.ts)

"declare let classes: {
  'class-a': string;
  'class-b': string;
};
export default classes;
```

It's mostly a matter of time, and I just haven't had the time to invest into rewriting the definitions to improve jump-to-definition... sorry!

**Comment by mrmckeb** [comment]
@cedeber perhaps you need to transform the classname? The below `tsconfig.json` example might work for you. The `camelCase` option gives both `axis_container` and `axisContainer`, or you can use `camelCaseOnly` to only get the latter of those :)
```json
{
  "compilerOptions": {
    "plugins": [
      {
        "name": "typescript-plugin-css-modules",
        "options": {
          "classnameTransform": "camelCase"
        }
      }
    ]
  }
}
```

**Comment by cedeber** [comment]
Yes, thanks, but unfortunately camelCase transformation doesn't work currently with Parcel 2. There is some issues with PostCSS apparently. But still, I prefer to keep the same name in the (S)CSS than in JS too, just to avoid confusion to other team mates who are not used to CSS modules "magic"

I'll try to figure out what's going on with simpler project.

**Comment by andykenward** [comment]
For Gatsby 3 and scss modules and you are using `gatsby-plugin-sass` you can set `namedExport: false`.

```
// gatsby.config.js
{
 resolve: `gatsby-plugin-sass`,
  options: {
   cssLoaderOptions: {
    modules: {
     namedExport: false,
    },
   },
  },
},
```

Then you can do

`import styles from 'styles.module.scss'` |

instead of 

`import * styles from 'styles.module.scss'`

**Comment by mrmckeb** [comment]
I'll close this off, but feel free to reopen in future.
